[Fix] Fix address uniqueness to use IO Type and address

eitch opened the pull request:
This should fix the issue where a PWM and GPIO use the same address as integer, but should be possible as they are different I/Os.

eitch committed:
[Fix] Fix address uniqueness to use IO Type and address

Refactored `Registry` to use `IOType` for managing addresses, ensuring better segregation of address spaces between IO types. Updated related methods and internal data structures accordingly.

FDelporte commented on pi4j-core/src/main/java/com/pi4j/registry/impl/DefaultRuntimeRegistry.java:
```diff
@@ -187,8 +197,9 @@ public synchronized boolean exists(String id) {
     @Override
-    public synchronized boolean exists(int address) {
-        return usedAddresses.contains(address);
+    public synchronized boolean exists(IOType ioType, int address) {
```
should this be unit tested?

eitch committed:
[Minor] Add support for shutdown by instance in `Registry` and `Context`

Introduced the ability to shutdown and unregister IO instances using the instance itself, complementing the existing ID-based method. Updated related methods, tests, and documentation to reflect the new functionality.

eitch commented:
@FDelporte i added tests

FDelporte commented:
> added tests
@eitch thanks, looks good!

FDelporte commented:
> looks good!

or not? ;-)

```
Error:    RegistryTest.testCreateMultipleSameAddress:102 » IOAlreadyExists IO instance [DIN-3] already exists in the Pi4J runtime context; unable to create a new instance using this reserved id.
Error:    RegistryTest.testShutdownAndRecreate:89 » IOAlreadyExists IO instance [DIN-3] already exists in the Pi4J runtime context; unable to create a new instance using this reserved id.
```

eitch committed:
[Fix] Ensure proper removal of IO instance during shutdown in `DefaultRuntimeRegistry`

eitch committed:
[Refactor] Replace `@BeforeAll` and `@AfterAll` with `@BeforeEach` and `@AfterEach` in `RegistryTest` for better test isolation.

eitch commented:
There was an issue with tests not cleaned up between each run. Is fixed now.

eitch commented:
@taartspi @FDelporte can we merge?

FDelporte commented:
> @taartspi @FDelporte can we merge?

LGTM

taartspi reviewed (approved):
Looks good
